### PR TITLE
test: improve resiliency of e2e tests on windows

### PR DESCRIPTION
**Pull request by sobolk**
#### Description of changes

This PR:
- makes sure that retry loop of e2e tests re-runs only failed tests not all tests
- make `yarn clean && yarn build-tests` possible (not leveraged in workspace build eventually due to other dependencies on full build but nice to have).

#### Description of how you validated changes

https://app.circleci.com/pipelines/github/aws-amplify/amplify-cli?branch=run-e2e%2Fsobkamil%2Fimprove-e2e-tests

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies
- [ ] [Pull request labels](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#labels) are added

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .circleci/local_publish_helpers.sh [hunk 1]
```diff
@@ -270,19 +270,12 @@ function setAwsAccountCredentials {
 function runE2eTest {
     FAILED_TEST_REGEX_FILE="./amplify-e2e-reports/amplify-e2e-failed-test.txt"
 
-    if [ -z "$FIRST_RUN" ] || [ "$FIRST_RUN" == "true" ]; then
-        startLocalRegistry "$(pwd)/.circleci/verdaccio.yaml"
-        setNpmRegistryUrlToLocal
-        changeNpmGlobalPath
-        cd $(pwd)/packages/amplify-e2e-tests
-    fi
```
Comment (sobolk) [inline .circleci/local_publish_helpers.sh:278]: This doesn't work on windows and has already been pulled up to circleci yaml for linux (and thus is redundant there).

File: nx.json [hunk 2]
```diff
@@ -31,6 +31,10 @@
       "inputs": ["default", "^production"],
       "dependsOn": ["^build"]
     },
+    "build-tests": {
+      "inputs": ["default", "^production"],
+      "dependsOn": ["build", "^build"]
+    },
```
Comment (sobolk) [inline nx.json:37]: this means that `build-tests` depends on packages `build` and dependencies `build` - this was missing to make `yarn build-tests` work at top level.

File: packages/amplify-cli-core/src/context/context-extensions.ts [hunk 3]
```diff
@@ -140,11 +140,11 @@ const contextFileSystem = {
...
-  read: (targetPath: string, encoding = 'utf8'): string => {
-    const result = fs.readFileSync(targetPath, encoding);
+  read: (targetPath: string, encoding: BufferEncoding = 'utf8'): string => {
+    const result = fs.readFileSync(targetPath, { encoding });
     return result;
   },
-  write: (targetPath: string, data: unknown): void => {
+  write: (targetPath: string, data: string | NodeJS.ArrayBufferView): void => {
     fs.ensureFileSync(targetPath);
     fs.writeFileSync(targetPath, data, 'utf-8');
```
Comment (sobolk) [inline packages/amplify-cli-core/src/context/context-extensions.ts:149]: type tweaks. if we run `yarn build-tests` from clean workspace `tsc` picks up more precise typing for `fs.read/writeFileSync` (i.e. it doesn't think it's `any`).
This then propagated around to callers.

File: packages/amplify-cli-core/src/jsonUtilities.ts [hunk 4]
```diff
@@ -124,7 +124,7 @@ export class JSONUtilities {
       minify?: boolean;
       orderedKeys?: boolean; // if true, will print object keys in alphabetical order
     },
-  ): string | undefined => {
+  ): string => {
```
Comment (sobolk) [inline packages/amplify-cli-core/src/jsonUtilities.ts:127]: there wasn't a codepath that would return undefined inside

File: .circleci/config.base.yml [hunk 5]
```diff
@@ -1303,7 +1303,7 @@ commands:
                   source $BASH_ENV
                   amplify version
                   cd packages/amplify-e2e-tests
-                  retry yarn run e2e --no-cache --maxWorkers=3 --forceExit $TEST_SUITE
+                  retry runE2eTest
```
Comment (sobolk) [inline .circleci/config.base.yml:1306]: this is the crux of the change. now windows e2e test will use same retry logic as linux.
I.e. when retry happens only tests that failed are re-attempted.

**Review by jhockett — APPROVED** [review]

**Review by sdstolworthy — APPROVED** [review]

**Pull request merged**